crates/hc_sandbox/src/run.rs

```diff
@@ -76,33 +85,89 @@ pub async fn run_async(
     }
     let config_path = write_config(sandbox_path.clone(), &config);
     let (tx_config, rx_config) = oneshot::channel();
-    let mut child = start_holochain(holochain_path, config_path, tx_config).await;
+    let (mut child, lair) = start_holochain(holochain_path, config_path, tx_config).await?;
     check_started(&mut child).await;
     let port = rx_config
         .await
         .expect("Failed to get admin port from conductor");
-    Ok((port, child))
+    Ok((port, child, lair))
 }
 
 async fn start_holochain(
     holochain_path: &Path,
     config_path: PathBuf,
     tx_config: oneshot::Sender<u16>,
-) -> Child {
+) -> anyhow::Result<(Child, Child)> {
+    use tokio::io::AsyncWriteExt;
+    let passphrase = holochain_util::pw::pw_get()?.read_lock().to_vec();
+
+    let mut lair_path = config_path.clone();
+    lair_path.pop();
+    lair_path.push("keystore");
+
+    tracing::info!("\n\n----\nstarting lair\n----\n\n");
+    let mut cmd = Command::new("lair-keystore");
+    cmd.arg("--lair-root")
+        .arg(lair_path)
+        .arg("server")
+        .arg("--piped")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::inherit())
+        .kill_on_drop(true);
+
+    msg!("{:?}", cmd);
+
+    let mut lair = cmd.spawn().expect("Failed to spawn lair-keystore");
+
+    let mut stdin = lair.stdin.take().unwrap();
+    stdin.write_all(&passphrase).await?;
+    stdin.shutdown().await?;
+    drop(stdin);
+
+    check_lair_running(lair.stdout.take().unwrap()).await;
+
     tracing::info!("\n\n----\nstarting holochain\n----\n\n");
     let mut cmd = Command::new(holochain_path);
     cmd.arg("--structured")
         // .env("RUST_LOG", "trace")
+        .arg("--piped")
         .arg("--config-path")
         .arg(config_path)
+        .stdin(Stdio::piped())
         .stdout(Stdio::piped())
         .stderr(Stdio::piped())
         .kill_on_drop(true);
 
+    msg!("{:?}", cmd);
+
     let mut holochain = cmd.spawn().expect("Failed to spawn holochain");
+
+    let mut stdin = holochain.stdin.take().unwrap();
+    stdin.write_all(&passphrase).await?;
+    stdin.shutdown().await?;
+    drop(stdin);
+
     // TODO: Allow redirecting output per conductor.
     spawn_output(&mut holochain, tx_config);
-    holochain
+    Ok((holochain, lair))
+}
+
+async fn check_lair_running(stdout: tokio::process::ChildStdout) {
+    let (s, r) = tokio::sync::oneshot::channel();
+    let mut s = Some(s);
+    tokio::task::spawn(async move {
+        let mut lines = BufReader::new(stdout).lines();
+        while let Ok(Some(line)) = lines.next_line().await {
+            println!("{}", line);
+            if line == "# lair-keystore running #" {
```

Comment:
any way to make this magic string something imported so that it won't get out of sync? or does it come from lair in a way that's not importable into this crate?

Reply:
we *could* publish it from lair, but i don't really want to make a version bump during the audit just for that... also the conductor command runner is doing something very similar in this same file.

Reply:
we could start by extracting it into a variable here and attach a todo to replace it by an import

Reply:
done